// @react-native-firebase/firestore/dist/module/web/RNFBFirestoreModule.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { setLogLevel, connectFirestoreEmulator, initializeFirestore, runTransaction, getApp, getFirestore, collection, collectionGroup, doc, getDoc, getDocs, getCount, getAggregate, count, average, sum, deleteDoc, setDoc, updateDoc, writeBatch, terminate } from '@react-native-firebase/app/dist/module/internal/web/firebaseFirestore';
import { guard, getWebError, emitEvent } from '@react-native-firebase/app/dist/module/internal/web/utils';
import { objectToWriteable, readableToObject, parseDocumentBatches } from "./convert.js";
import { buildQuery } from "./query.js";
import { executeWebSdkPipeline } from "./pipelines/pipeline.js";
function rejectWithCodeAndMessage(code, message) {
  return Promise.reject(getWebError({
    code,
    message
  }));
}
function documentSnapshotToObject(snapshot) {
  const exists = snapshot.exists();
  const out = {
    metadata: [false, false],
    path: snapshot.ref.path,
    exists
  };
  if (exists) {
    out.data = objectToWriteable(snapshot.data() ?? {});
  }
  return out;
}
function querySnapshotToObject(snapshot) {
  return {
    source: 'get',
    excludesMetadataChanges: true,
    changes: [],
    metadata: [false, false],
    documents: snapshot.docs.map(documentSnapshotToObject)
  };
}
const emulatorForApp = {};
const firestoreInstances = {};
const appInstances = {};
const transactionHandler = {};
const transactionBuffer = {};
function getCachedAppInstance(appName) {
  return appInstances[appName] ??= getApp(appName);
}
function createFirestoreKey(appName, databaseId) {
  return `${appName}:${databaseId}`;
}
function getCachedFirestoreInstance(appName, databaseId) {
  const firestoreKey = createFirestoreKey(appName, databaseId);
  let instance = firestoreInstances[firestoreKey];
  if (!instance) {
    instance = getFirestore(getCachedAppInstance(appName), databaseId);
    if (emulatorForApp[firestoreKey]) {
      connectFirestoreEmulator(instance, emulatorForApp[firestoreKey].host, emulatorForApp[firestoreKey].port);
    }
    firestoreInstances[firestoreKey] = instance;
  }
  return instance;
}
export default {
  async setLogLevel(logLevel) {
    if (logLevel === 'debug' || logLevel === 'error') {
      setLogLevel(logLevel);
    } else {
      setLogLevel('silent');
    }
  },
  loadBundle() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  clearPersistence() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  async waitForPendingWrites() {
    return null;
  },
  disableNetwork() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  enableNetwork() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  addSnapshotsInSync() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  removeSnapshotsInSync() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  useEmulator(appName, databaseId, host, port) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      connectFirestoreEmulator(firestore, host, port);
      const firestoreKey = createFirestoreKey(appName, databaseId);
      emulatorForApp[firestoreKey] = {
        host,
        port
      };
      return null;
    });
  },
  settings(appName, databaseId, settings) {
    return guard(async () => {
      const instance = initializeFirestore(getCachedAppInstance(appName), settings, databaseId);
      firestoreInstances[appName] = instance;
      return null;
    });
  },
  terminate(appName, databaseId) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      await terminate(firestore);
      return null;
    });
  },
  namedQueryOnSnapshot() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  collectionOnSnapshot() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  collectionOffSnapshot() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  namedQueryGet() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  collectionCount(appName, databaseId, path, type, filters, orders, options) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const queryRef = type === 'collectionGroup' ? collectionGroup(firestore, path) : collection(firestore, path);
      const query = buildQuery(queryRef, filters, orders, options);
      const snapshot = await getCount(query);
      return {
        count: snapshot.data().count
      };
    });
  },
  aggregateQuery(appName, databaseId, path, type, filters, orders, options, aggregateQueries) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const queryRef = type === 'collectionGroup' ? collectionGroup(firestore, path) : collection(firestore, path);
      const query = buildQuery(queryRef, filters, orders, options);
      const aggregateSpec = {};
      for (let i = 0; i < aggregateQueries.length; i++) {
        const aggregateQuery = aggregateQueries[i];
        const {
          aggregateType,
          field,
          key
        } = aggregateQuery;
        switch (aggregateType) {
          case 'count':
            aggregateSpec[key] = count();
            break;
          case 'avg':
            aggregateSpec[key] = average(field);
            break;
          case 'sum':
            aggregateSpec[key] = sum(field);
            break;
        }
      }
      const result = await getAggregate(query, aggregateSpec);
      const data = result.data();
      const response = {};
      for (let i = 0; i < aggregateQueries.length; i++) {
        const aggregateQuery = aggregateQueries[i];
        response[aggregateQuery.key] = data[aggregateQuery.key];
      }
      return response;
    });
  },
  pipelineExecute(appName, databaseId, pipeline, options) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      return executeWebSdkPipeline(firestore, pipeline, options);
    });
  },
  collectionGet(appName, databaseId, path, type, filters, orders, options, getOptions) {
    if (getOptions?.source === 'cache') {
      return rejectWithCodeAndMessage('unsupported', 'The source cache is not supported in the lite SDK.');
    }
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const queryRef = type === 'collectionGroup' ? collectionGroup(firestore, path) : collection(firestore, path);
      const query = buildQuery(queryRef, filters, orders, options);
      const snapshot = await getDocs(query);
      return querySnapshotToObject(snapshot);
    });
  },
  documentOnSnapshot() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  documentOffSnapshot() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  persistenceCacheIndexManager() {
    return rejectWithCodeAndMessage('unsupported', 'Not supported in the lite SDK.');
  },
  documentGet(appName, databaseId, path, getOptions) {
    return guard(async () => {
      if (getOptions?.source === 'cache') {
        return rejectWithCodeAndMessage('unsupported', 'The source cache is not supported in the lite SDK.');
      }
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const ref = doc(firestore, path);
      const snapshot = await getDoc(ref);
      return documentSnapshotToObject(snapshot);
    });
  },
  documentDelete(appName, databaseId, path) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const ref = doc(firestore, path);
      await deleteDoc(ref);
      return null;
    });
  },
  documentSet(appName, databaseId, path, data, options) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const ref = doc(firestore, path);
      const setOptions = {};
      if (options && 'merge' in options) {
        setOptions.merge = options.merge;
      } else if (options && 'mergeFields' in options) {
        setOptions.mergeFields = options.mergeFields;
      }
      await setDoc(ref, readableToObject(firestore, data), setOptions);
    });
  },
  documentUpdate(appName, databaseId, path, data) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const ref = doc(firestore, path);
      await updateDoc(ref, readableToObject(firestore, data));
    });
  },
  documentBatch(appName, databaseId, writes) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const batch = writeBatch(firestore);
      const writesArray = parseDocumentBatches(firestore, writes);
      for (const parsed of writesArray) {
        const {
          type,
          path
        } = parsed;
        const ref = doc(firestore, path);
        switch (type) {
          case 'DELETE':
            batch.delete(ref);
            break;
          case 'UPDATE':
            batch.update(ref, parsed.data);
            break;
          case 'SET':
            {
              const options = parsed.options ?? {};
              const setOptions = {};
              if ('merge' in options) {
                setOptions.merge = options.merge;
              } else if ('mergeFields' in options) {
                setOptions.mergeFields = options.mergeFields;
              }
              batch.set(ref, parsed.data, setOptions);
              break;
            }
        }
      }
      await batch.commit();
    });
  },
  transactionGetDocument(appName, databaseId, transactionId, path) {
    if (!transactionHandler[transactionId]) {
      return rejectWithCodeAndMessage('internal-error', 'An internal error occurred whilst attempting to find a native transaction by id.');
    }
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      const docRef = doc(firestore, path);
      const tsx = transactionHandler[transactionId];
      const snapshot = await tsx.get(docRef);
      return documentSnapshotToObject(snapshot);
    });
  },
  transactionDispose(_appName, _databaseId, transactionId) {
    delete transactionHandler[transactionId];
  },
  transactionApplyBuffer(_appName, _databaseId, transactionId, commandBuffer) {
    if (transactionHandler[transactionId]) {
      transactionBuffer[transactionId] = commandBuffer;
    }
  },
  transactionBegin(appName, databaseId, transactionId) {
    return guard(async () => {
      const firestore = getCachedFirestoreInstance(appName, databaseId);
      try {
        await runTransaction(firestore, async tsx => {
          transactionHandler[transactionId] = tsx;
          emitEvent('firestore_transaction_event', {
            eventName: 'firestore_transaction_event',
            body: {
              type: 'update'
            },
            appName,
            databaseId,
            listenerId: transactionId
          });
          const getBuffer = () => transactionBuffer[transactionId];
          const buffer = await new Promise(resolve => {
            const interval = setInterval(() => {
              const b = getBuffer();
              if (b) {
                clearInterval(interval);
                resolve(b);
              }
            }, 100);
          });
          for (const serialized of buffer) {
            const {
              path: docPath,
              type,
              data
            } = serialized;
            const docRef = doc(firestore, docPath);
            switch (type) {
              case 'DELETE':
                tsx.delete(docRef);
                break;
              case 'UPDATE':
                tsx.update(docRef, readableToObject(firestore, data ?? {}));
                break;
              case 'SET':
                {
                  const options = serialized.options ?? {};
                  const setOptions = {};
                  if ('merge' in options) {
                    setOptions.merge = options.merge;
                  } else if ('mergeFields' in options) {
                    setOptions.mergeFields = options.mergeFields;
                  }
                  tsx.set(docRef, readableToObject(firestore, data ?? {}), setOptions);
                  break;
                }
            }
          }
        });
        emitEvent('firestore_transaction_event', {
          eventName: 'firestore_transaction_event',
          body: {
            type: 'complete'
          },
          appName,
          databaseId,
          listenerId: transactionId
        });
      } catch (e) {
        emitEvent('firestore_transaction_event', {
          eventName: 'firestore_transaction_event',
          body: {
            type: 'error',
            error: getWebError(e)
          },
          appName,
          databaseId,
          listenerId: transactionId
        });
      }
    });
  }
};
