// @react-native-firebase/firestore/dist/module/FirestoreFilter.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { isArray, isNull, isString, isUndefined } from '@react-native-firebase/app/dist/module/common';
import { fromDotSeparatedString } from "./FieldPath.js";
import { OPERATORS } from "./FirestoreQueryModifiers.js";
import { generateNativeData } from "./utils/serialize.js";
const AND_QUERY = 'AND';
const OR_QUERY = 'OR';

/** Value allowed in a filter (primitive, object, or array for in/array-contains-any). */

export function Filter(fieldPath, operator, value) {
  return new _Filter(fieldPath, operator, value);
}
export class _Filter {
  constructor(fieldPath, operator, value, filterOperator, queries) {
    if (filterOperator !== undefined && [AND_QUERY, OR_QUERY].includes(filterOperator)) {
      this.operator = filterOperator;
      this.queries = queries;
      return;
    }
    this.fieldPath = fieldPath ?? undefined;
    this.operator = operator;
    // Preserve null for == / != (required for mapFieldQuery; do not coerce to undefined)
    this.value = value;
  }
  _toMap() {
    if ([AND_QUERY, OR_QUERY].includes(this.operator)) {
      return {
        operator: this.operator,
        queries: this.queries.map(query => query._toMap())
      };
    }
    return {
      fieldPath: this.fieldPath,
      operator: this.operator,
      value: this.value
    };
  }
}
Filter.and = function and(...queries) {
  if (queries.length > 10 || queries.length < 1) {
    throw new Error(`Expected 1-10 instances of Filter, but got ${queries.length} Filters`);
  }
  const validateFilters = queries.every(filter => filter instanceof _Filter);
  if (!validateFilters) {
    throw new Error('Expected every argument to be an instance of Filter');
  }
  return new _Filter(null, null, null, AND_QUERY, queries);
};

/** Serialized field filter or composite (AND/OR) for generateFilters output. */

function hasOrOperator(obj) {
  return obj.operator === OR_QUERY || Array.isArray(obj.queries) && obj.queries.some(query => hasOrOperator(query));
}
Filter.or = function or(...queries) {
  if (queries.length > 10 || queries.length < 1) {
    throw new Error(`Expected 1-10 instances of Filter, but got ${queries.length} Filters`);
  }
  const validateFilters = queries.every(filter => filter instanceof _Filter);
  if (!validateFilters) {
    throw new Error('Expected every argument to be an instance of Filter');
  }
  if (queries.some(hasOrOperator)) {
    throw new Error('OR Filters with nested OR Filters are not supported');
  }
  return new _Filter(null, null, null, OR_QUERY, queries);
};

/** Modifiers object passed to generateFilters (e.g. QueryModifiers). */

function mapFieldQuery(filterMap, modifiers) {
  const {
    fieldPath,
    operator,
    value,
    queries
  } = filterMap;
  if (operator === AND_QUERY || operator === OR_QUERY) {
    return {
      operator,
      queries: (queries ?? []).map(q => mapFieldQuery(q, modifiers))
    };
  }
  let path;
  if (isString(fieldPath)) {
    try {
      path = fromDotSeparatedString(fieldPath);
    } catch (e) {
      throw new Error(`first argument of Filter(*,_ , _) 'fieldPath' ${e.message}.`);
    }
  } else if (fieldPath !== undefined) {
    path = fieldPath;
  } else {
    throw new Error("first argument of Filter(*,_ , _) 'fieldPath' is required.");
  }
  if (!modifiers.isValidOperator(operator)) {
    throw new Error("second argument of Filter(*,_ , _) 'opStr' is invalid. Expected one of '==', '>', '>=', '<', '<=', '!=', 'array-contains', 'not-in', 'array-contains-any' or 'in'.");
  }
  if (isUndefined(value)) {
    throw new Error("third argument of Filter(*,_ , _) 'value' argument expected.");
  }
  if (isNull(value) && !modifiers.isEqualOperator(operator) && !modifiers.isNotEqualOperator(operator)) {
    throw new Error("third argument of Filter(*,_ , _) 'value' is invalid. You can only perform equals comparisons on null");
  }
  if (modifiers.isInOperator(operator)) {
    if (!isArray(value) || !value.length) {
      throw new Error(`third argument of Filter(*,_ , _) 'value' is invalid. A non-empty array is required for '${String(operator)}' filters.`);
    }
    if (value.length > 10) {
      throw new Error(`third argument of Filter(*,_ , _) 'value' is invalid. '${String(operator)}' filters support a maximum of 10 elements in the value array.`);
    }
  }
  return {
    fieldPath: path,
    operator: OPERATORS[operator],
    value: generateNativeData(value, true)
  };
}
export function generateFilters(filter, modifiers) {
  const filterMap = filter._toMap();
  const queries = (filterMap.queries ?? []).map(q => mapFieldQuery(q, modifiers));
  return {
    operator: filterMap.operator ?? '',
    queries
  };
}
