// @react-native-firebase/firestore/dist/module/FirestoreDocumentSnapshot.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { isString } from '@react-native-firebase/app/dist/module/common';
import DocumentReference, { provideDocumentSnapshotClass } from "./FirestoreDocumentReference.js";
import FieldPath, { fromDotSeparatedString } from "./FieldPath.js";
import FirestorePath from "./FirestorePath.js";
import SnapshotMetadata from "./FirestoreSnapshotMetadata.js";
import { extractFieldPathData } from "./utils/index.js";
import { parseNativeMap } from "./utils/serialize.js";
export default class DocumentSnapshot {
  constructor(firestore, nativeData, converter) {
    this._firestore = firestore;
    this._nativeData = nativeData;
    this._converter = converter;
    this._data = parseNativeMap(firestore, nativeData.data);
    this._metadata = new SnapshotMetadata(nativeData.metadata ?? [false, false]);
    this._ref = new DocumentReference(firestore, FirestorePath.fromName(nativeData.path), this._converter);
    this._exists = nativeData.exists ?? false;
  }
  get id() {
    return this._ref.id;
  }
  get metadata() {
    return this._metadata;
  }
  get ref() {
    return this._ref;
  }
  exists() {
    return this._exists;
  }
  data(options) {
    if (this._converter) {
      try {
        return this._converter.fromFirestore(new DocumentSnapshot(this._firestore, this._nativeData, null), options);
      } catch (e) {
        throw new Error(`firebase.firestore() DocumentSnapshot.data(*) 'withConverter.fromFirestore' threw an error: ${e.message}.`);
      }
    }
    return this._data;
  }
  get(fieldPath, _options) {
    if (!isString(fieldPath) && !(fieldPath instanceof FieldPath)) {
      throw new Error("firebase.firestore() DocumentSnapshot.get(*) 'fieldPath' expected type string or FieldPath.");
    }
    let path;
    if (isString(fieldPath)) {
      try {
        path = fromDotSeparatedString(fieldPath);
      } catch (e) {
        throw new Error(`firebase.firestore() DocumentSnapshot.get(*) 'fieldPath' ${e.message}.`);
      }
    } else {
      path = fieldPath;
    }
    return extractFieldPathData(this._data, path._segments);
  }
  isEqual(other) {
    if (!(other instanceof DocumentSnapshot)) {
      throw new Error("firebase.firestore() DocumentSnapshot.isEqual(*) 'other' expected a DocumentSnapshot instance.");
    }
    if (this.exists() !== other.exists() || !this.metadata.isEqual(other.metadata) || !this.ref.isEqual(other.ref)) {
      return false;
    }
    const thisData = JSON.stringify(this.data());
    const otherData = JSON.stringify(other.data());
    return thisData === otherData;
  }
}
provideDocumentSnapshotClass(DocumentSnapshot);
