// @react-native-firebase/firestore/dist/module/FirestoreQuerySnapshot.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { createDeprecationProxy, isBoolean, isFunction, isObject, isUndefined } from '@react-native-firebase/app/dist/module/common';
import DocumentChange from "./FirestoreDocumentChange.js";
import DocumentSnapshot from "./FirestoreDocumentSnapshot.js";
import SnapshotMetadata from "./FirestoreSnapshotMetadata.js";
export default class QuerySnapshot {
  constructor(firestore, query, nativeData, converter) {
    this._query = query;
    this._source = nativeData.source;
    this._excludesMetadataChanges = nativeData.excludesMetadataChanges;
    this._changes = nativeData.changes.map(c => new DocumentChange(firestore, c, converter));
    this._docs = nativeData.documents.map(doc => createDeprecationProxy(new DocumentSnapshot(firestore, doc, converter)));
    this._metadata = new SnapshotMetadata(nativeData.metadata ?? [false, false]);
  }
  get docs() {
    return this._docs;
  }
  get empty() {
    return this._docs.length === 0;
  }
  get metadata() {
    return this._metadata;
  }
  get query() {
    return this._query;
  }
  get size() {
    return this._docs.length;
  }
  docChanges(options) {
    if (!isUndefined(options) && !isObject(options)) {
      throw new Error("firebase.firestore() QuerySnapshot.docChanges(*) 'options' expected an object.");
    }
    let includeMetaDataChanges = false;
    if (options) {
      if (!isBoolean(options.includeMetadataChanges)) {
        throw new Error("firebase.firestore() QuerySnapshot.docChanges(*) 'options.includeMetadataChanges' expected a boolean.");
      }
      includeMetaDataChanges = options.includeMetadataChanges ?? false;
    }
    if (this._source === 'get') {
      return this._changes;
    }
    if (includeMetaDataChanges && this._excludesMetadataChanges) {
      throw new Error('firebase.firestore() QuerySnapshot.docChanges() To include metadata changes with your document changes, you must also pass { includeMetadataChanges:true } to onSnapshot().');
    }
    return this._changes.filter($ => {
      if (!includeMetaDataChanges) {
        return $._isMetadataChange === false;
      }
      return true;
    });
  }
  forEach(callback, thisArg) {
    if (!isFunction(callback)) {
      throw new Error("firebase.firestore() QuerySnapshot.forEach(*) 'callback' expected a function.");
    }
    const cb = thisArg ? callback.bind(thisArg) : callback;
    for (let i = 0; i < this._docs.length; i++) {
      cb(this._docs[i], i);
    }
  }
  isEqual(other, ...args) {
    if (!(other instanceof QuerySnapshot)) {
      throw new Error("firebase.firestore() QuerySnapshot.isEqual(*) 'other' expected a QuerySnapshot instance.");
    }
    if (this.empty !== other.empty || this.size !== other.size || !this.metadata.isEqual(other.metadata)) {
      return false;
    }
    for (let i = 0; i < this.docs.length; i++) {
      const thisDoc = this.docs[i];
      const otherDoc = other.docs[i];
      if (!thisDoc.isEqual(otherDoc, ...args)) {
        return false;
      }
    }
    return true;
  }
}
