// @react-native-firebase/firestore/dist/module/modular.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { getApp, setLogLevel as appSetLogLevel } from '@react-native-firebase/app';
import { isObject, MODULAR_DEPRECATION_ARG } from '@react-native-firebase/app/dist/module/common';
import { AggregateField, fieldPathFromArgument, AggregateQuerySnapshot } from "./FirestoreAggregate.js";
import QueryImpl from "./FirestoreQuery.js";
import { LoadBundleTask } from "./LoadBundleTask.js";
export { AggregateField, AggregateQuerySnapshot } from "./FirestoreAggregate.js";
export const CACHE_SIZE_UNLIMITED = -1;
const PIPELINE_RUNTIME_INSTALLER_SYMBOL = Symbol.for('RNFBFirestorePipelineRuntimeInstaller');
export function getFirestore(appOrDatabaseId, databaseId) {
  const app = name => getApp(name);
  let firestore;
  if (typeof appOrDatabaseId === 'string') {
    firestore = app().firestore(appOrDatabaseId);
  } else if (appOrDatabaseId) {
    if (databaseId) {
      firestore = app(appOrDatabaseId.name).firestore(databaseId);
    } else {
      firestore = app(appOrDatabaseId.name).firestore();
    }
  } else if (databaseId) {
    firestore = app().firestore(databaseId);
  } else {
    firestore = app().firestore();
  }
  const runtimeGlobal = globalThis;
  const installPipelineRuntime = runtimeGlobal[PIPELINE_RUNTIME_INSTALLER_SYMBOL];
  if (typeof installPipelineRuntime === 'function') {
    try {
      installPipelineRuntime(firestore);
    } catch {
      // Avoid changing getFirestore behavior if optional pipeline runtime install fails.
    }
  }
  return firestore;
}
export function doc(parent, path, ...pathSegments) {
  let resolvedPath = path;
  if (pathSegments.length) {
    resolvedPath = (resolvedPath ?? '') + '/' + pathSegments.map(segment => segment.replace(/^\/|\/$/g, '')).join('/');
  }
  return parent.doc.call(parent, resolvedPath, MODULAR_DEPRECATION_ARG);
}
export function collection(parent, path, ...pathSegments) {
  let resolvedPath = path;
  if (pathSegments.length) {
    resolvedPath = `${resolvedPath}/${pathSegments.map(segment => segment.replace(/^\/|\/$/g, '')).join('/')}`;
  }
  return parent.collection.call(parent, resolvedPath, MODULAR_DEPRECATION_ARG);
}
export function refEqual(left, right) {
  return left.isEqual.call(left, right, MODULAR_DEPRECATION_ARG);
}
export function collectionGroup(firestore, collectionId) {
  return firestore.collectionGroup.call(firestore, collectionId, MODULAR_DEPRECATION_ARG);
}
let snapshotInSyncListenerId = 0;
export function onSnapshotsInSync(firestore, callback) {
  const listenerId = snapshotInSyncListenerId++;
  const syncFirestore = firestore;
  syncFirestore.native.addSnapshotsInSync(listenerId);
  const subscription = syncFirestore.emitter.addListener(syncFirestore.eventNameForApp(`firestore_snapshots_in_sync_event:${listenerId}`), () => {
    if (typeof callback === 'function') {
      callback();
    } else {
      callback.next?.();
    }
  });
  return () => {
    subscription.remove();
    syncFirestore.native.removeSnapshotsInSync(listenerId);
  };
}
export function setDoc(reference, data, options) {
  return reference.set.call(reference, data, options, MODULAR_DEPRECATION_ARG);
}
export function updateDoc(reference, fieldOrUpdateData, value, ...moreFieldsAndValues) {
  const ref = reference;
  if (!fieldOrUpdateData) {
    return ref.update.call(reference, MODULAR_DEPRECATION_ARG);
  }
  if (!value) {
    return ref.update.call(reference, fieldOrUpdateData, MODULAR_DEPRECATION_ARG);
  }
  if (!Array.isArray(moreFieldsAndValues)) {
    return ref.update.call(reference, fieldOrUpdateData, value, MODULAR_DEPRECATION_ARG);
  }
  return ref.update.call(reference, fieldOrUpdateData, value, ...moreFieldsAndValues, MODULAR_DEPRECATION_ARG);
}
export function addDoc(reference, data) {
  return reference.add.call(reference, data, MODULAR_DEPRECATION_ARG);
}
export function enableNetwork(firestore) {
  return firestore.enableNetwork.call(firestore, MODULAR_DEPRECATION_ARG);
}
export function disableNetwork(firestore) {
  return firestore.disableNetwork.call(firestore, MODULAR_DEPRECATION_ARG);
}
export function clearPersistence(firestore) {
  // this will call deprecation warning as it isn't part of firebase-js-sdk API
  return firestore.clearPersistence();
}
export function clearIndexedDbPersistence(firestore) {
  return firestore.clearPersistence.call(firestore, MODULAR_DEPRECATION_ARG);
}
export function terminate(firestore) {
  return firestore.terminate.call(firestore, MODULAR_DEPRECATION_ARG);
}
export function waitForPendingWrites(firestore) {
  return firestore.waitForPendingWrites.call(firestore, MODULAR_DEPRECATION_ARG);
}
export async function initializeFirestore(app, settings, databaseId) {
  const firebase = getApp(app.name);
  const firestore = firebase.firestore(databaseId);
  await firestore.settings.call(firestore, settings, MODULAR_DEPRECATION_ARG);
  return firestore;
}
export function connectFirestoreEmulator(firestore, host, port, options) {
  return firestore.useEmulator.call(firestore, host, port, options, MODULAR_DEPRECATION_ARG);
}
export function setLogLevel(logLevel) {
  return appSetLogLevel(logLevel);
}
export function runTransaction(firestore, updateFunction) {
  return firestore.runTransaction.call(firestore, updateFunction, MODULAR_DEPRECATION_ARG);
}
export function getCountFromServer(query) {
  return query.count.call(query, MODULAR_DEPRECATION_ARG).get();
}
export function getAggregateFromServer(query, aggregateSpec) {
  if (!(query instanceof QueryImpl)) {
    throw new Error('`getAggregateFromServer(*, aggregateSpec)` `query` must be an instance of `FirestoreQuery`');
  }
  if (!isObject(aggregateSpec)) {
    throw new Error('`getAggregateFromServer(query, *)` `aggregateSpec` must be an object');
  }
  const containsOneAggregateField = Object.values(aggregateSpec).find(value => value instanceof AggregateField);
  if (!containsOneAggregateField) {
    throw new Error('`getAggregateFromServer(query, *)` `aggregateSpec` must contain at least one `AggregateField`');
  }
  const aggregateQueries = [];
  for (const key in aggregateSpec) {
    if (!Object.prototype.hasOwnProperty.call(aggregateSpec, key)) {
      continue;
    }
    const aggregateField = aggregateSpec[key];
    if (!(aggregateField instanceof AggregateField)) {
      continue;
    }
    switch (aggregateField.aggregateType) {
      case 'avg':
      case 'sum':
      case 'count':
        aggregateQueries.push({
          aggregateType: aggregateField.aggregateType,
          field: aggregateField._internalFieldPath === undefined ? null : aggregateField._internalFieldPath._toPath(),
          key
        });
        break;
      default:
        throw new Error(`'AggregateField' has an an unknown 'AggregateType' : ${aggregateField.aggregateType}`);
    }
  }
  const queryWithInternals = query;
  return queryWithInternals._firestore.native.aggregateQuery(queryWithInternals._collectionPath.relativeName, queryWithInternals._modifiers.type, queryWithInternals._modifiers.filters, queryWithInternals._modifiers.orders, queryWithInternals._modifiers.options, aggregateQueries).then(data => new AggregateQuerySnapshot(query, data, false));
}
export function sum(field) {
  return new AggregateField('sum', fieldPathFromArgument(field));
}
export function average(field) {
  return new AggregateField('avg', fieldPathFromArgument(field));
}
export function count() {
  return new AggregateField('count');
}
export function loadBundle(firestore, bundleData) {
  const task = new LoadBundleTask();
  firestore.loadBundle.call(firestore, bundleData, MODULAR_DEPRECATION_ARG).then(progress => task._completeWith(progress)).catch(error => task._failWith(error));
  return task;
}
export function namedQuery(firestore, name) {
  return Promise.resolve(firestore.namedQuery.call(firestore, name, MODULAR_DEPRECATION_ARG));
}
export function writeBatch(firestore) {
  return firestore.batch.call(firestore, MODULAR_DEPRECATION_ARG);
}
export function getPersistentCacheIndexManager(firestore) {
  return firestore.persistentCacheIndexManager.call(firestore, MODULAR_DEPRECATION_ARG);
}
export function enablePersistentCacheIndexAutoCreation(indexManager) {
  return indexManager.enableIndexAutoCreation.call(indexManager, MODULAR_DEPRECATION_ARG);
}
export function disablePersistentCacheIndexAutoCreation(indexManager) {
  return indexManager.disableIndexAutoCreation.call(indexManager, MODULAR_DEPRECATION_ARG);
}
export function deleteAllPersistentCacheIndexes(indexManager) {
  return indexManager.deleteAllIndexes.call(indexManager, MODULAR_DEPRECATION_ARG);
}
export * from "./modular/query.js";
export * from "./modular/snapshot.js";
export * from "./modular/Bytes.js";
export * from "./modular/FieldPath.js";
export * from "./modular/FieldValue.js";
export * from "./modular/GeoPoint.js";
export * from "./modular/Timestamp.js";
export * from "./modular/VectorValue.js";
export { LoadBundleTask } from "./LoadBundleTask.js";
export { default as Transaction } from "./FirestoreTransaction.js";
export { default as WriteBatch } from "./FirestoreWriteBatch.js";
export { Filter } from "./FirestoreFilter.js";
