// @react-native-firebase/firestore/dist/module/web/pipelines/pipeline_bridge_factory.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { collection, collectionGroup } from '@react-native-firebase/app/dist/module/internal/web/firebaseFirestore';
import { revivePipelineValue } from "./pipeline_node_builder.js";
import { buildQuery } from "../query.js";
const PIPELINE_RUNTIME_IMPORT_SUFFIX = ' Import "@react-native-firebase/firestore/pipelines" before using pipelines so the pipeline runtime is installed.';
function createPipelineRuntimeImportError(message) {
  return new Error(message + PIPELINE_RUNTIME_IMPORT_SUFFIX);
}
function isRecord(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}
function buildCollectionSourcePipeline(pipelineSource, source) {
  if (typeof pipelineSource.collection !== 'function') {
    return undefined;
  }
  return pipelineSource.collection({
    collection: source.path,
    ...(isRecord(source.rawOptions) && source.rawOptions ? {
      rawOptions: source.rawOptions
    } : {})
  });
}
function buildCollectionGroupSourcePipeline(pipelineSource, source) {
  if (typeof pipelineSource.collectionGroup !== 'function') {
    return undefined;
  }
  return pipelineSource.collectionGroup({
    collectionId: source.collectionId,
    ...(isRecord(source.rawOptions) && source.rawOptions ? {
      rawOptions: source.rawOptions
    } : {})
  });
}
function buildDatabaseSourcePipeline(pipelineSource, source) {
  if (typeof pipelineSource.database !== 'function') {
    return undefined;
  }
  return pipelineSource.database(isRecord(source.rawOptions) && source.rawOptions ? {
    rawOptions: source.rawOptions
  } : {});
}
function buildDocumentsSourcePipeline(pipelineSource, source) {
  if (typeof pipelineSource.documents !== 'function') {
    return undefined;
  }
  return pipelineSource.documents({
    docs: source.documents,
    ...(isRecord(source.rawOptions) && source.rawOptions ? {
      rawOptions: source.rawOptions
    } : {})
  });
}
function buildQuerySourcePipeline(firestore, pipelineSource, source) {
  if (typeof pipelineSource.createFrom !== 'function') {
    throw createPipelineRuntimeImportError('pipelineExecute() expected pipeline source to support createFrom(query).');
  }
  const queryRef = source.queryType === 'collectionGroup' ? collectionGroup(firestore, source.path) : collection(firestore, source.path);
  const query = buildQuery(queryRef, source.filters, source.orders, source.options);
  return pipelineSource.createFrom(query);
}
function normalizeWebDistanceMeasure(value, fieldName) {
  if (typeof value !== 'string') {
    throw new Error(`pipelineExecute() expected ${fieldName} to be a string.`);
  }
  const normalized = value.trim().replace(/-/g, '_').replace(/ /g, '_').toUpperCase();
  switch (normalized) {
    case 'COSINE':
      return 'cosine';
    case 'EUCLIDEAN':
      return 'euclidean';
    case 'DOT_PRODUCT':
    case 'DOTPRODUCT':
      return 'dot_product';
    default:
      throw new Error(`pipelineExecute() expected ${fieldName} to be one of euclidean, cosine, or dot_product.`);
  }
}
function normalizeWebFindNearestOptions(options) {
  if (!Object.prototype.hasOwnProperty.call(options, 'distanceMeasure')) {
    return options;
  }
  return {
    ...options,
    distanceMeasure: normalizeWebDistanceMeasure(options.distanceMeasure, 'stage.options.distanceMeasure')
  };
}
function getPipelineStageMethod(current, stageName) {
  const method = current?.[stageName];
  if (typeof method !== 'function') {
    throw new Error(`Pipeline stage "${stageName}" is not supported by the current web SDK.`);
  }
  return method;
}
function applyPipelineStage(current, stage, nestedUnionPipeline) {
  if (stage.stage === 'union') {
    return getPipelineStageMethod(current, stage.stage).call(current, {
      other: nestedUnionPipeline
    });
  }
  const revivedStageOptions = revivePipelineValue(stage.options);
  const stageArgs = isRecord(revivedStageOptions) ? revivedStageOptions : stage.options;
  const method = getPipelineStageMethod(current, stage.stage);
  switch (stage.stage) {
    case 'where':
      if (isRecord(stageArgs.condition) && typeof stageArgs.condition.operator === 'string') {
        const op = stageArgs.condition.operator.toUpperCase();
        const isArrayOperator = op === 'IN' || op === 'NOT_IN' || op === 'ARRAY_CONTAINS_ANY' || op === 'ARRAY_CONTAINS_ALL';
        if (isArrayOperator && !Array.isArray(stageArgs.condition.value)) {
          throw new Error('invalid argument');
        }
      }
      return stageArgs.condition !== undefined ? method.call(current, stageArgs.condition) : method.call(current, stageArgs);
    case 'select':
      return Array.isArray(stageArgs.selections) ? method.call(current, ...stageArgs.selections) : method.call(current, stageArgs);
    case 'addFields':
      return Array.isArray(stageArgs.fields) ? method.call(current, ...stageArgs.fields) : method.call(current, stageArgs);
    case 'removeFields':
      return Array.isArray(stageArgs.fields) ? method.call(current, ...stageArgs.fields) : method.call(current, stageArgs);
    case 'sort':
      return Array.isArray(stageArgs.orderings) ? method.call(current, ...stageArgs.orderings) : method.call(current, stageArgs);
    case 'limit':
      return typeof stageArgs.limit === 'number' ? method.call(current, stageArgs.limit) : method.call(current, stageArgs);
    case 'offset':
      return typeof stageArgs.offset === 'number' ? method.call(current, stageArgs.offset) : method.call(current, stageArgs);
    case 'distinct':
      return Array.isArray(stageArgs.groups) ? method.call(current, ...stageArgs.groups) : method.call(current, stageArgs);
    case 'aggregate':
      return method.call(current, stageArgs);
    case 'replaceWith':
      return stageArgs.map !== undefined ? method.call(current, stageArgs.map) : method.call(current, stageArgs);
    case 'sample':
      return typeof stageArgs.documents === 'number' ? method.call(current, stageArgs.documents) : method.call(current, stageArgs);
    case 'unnest':
      return stageArgs.selectable !== undefined ? method.call(current, stageArgs.selectable, stageArgs.indexField) : method.call(current, stageArgs);
    case 'rawStage':
      return typeof stageArgs.name === 'string' ? method.call(current, stageArgs.name, stageArgs.params, isRecord(stageArgs.options) ? stageArgs.options : undefined) : method.call(current, stageArgs);
    case 'findNearest':
      return method.call(current, normalizeWebFindNearestOptions(stageArgs));
    default:
      return method.call(current, stageArgs);
  }
}
function buildSourcePipeline(firestore, pipelineFactory, pipeline) {
  const pipelineSource = pipelineFactory.call(firestore);
  const source = pipeline.source;
  let currentPipeline;
  switch (source.source) {
    case 'collection':
      currentPipeline = buildCollectionSourcePipeline(pipelineSource, source);
      break;
    case 'collectionGroup':
      currentPipeline = buildCollectionGroupSourcePipeline(pipelineSource, source);
      break;
    case 'database':
      currentPipeline = buildDatabaseSourcePipeline(pipelineSource, source);
      break;
    case 'documents':
      currentPipeline = buildDocumentsSourcePipeline(pipelineSource, source);
      break;
    case 'query':
      currentPipeline = buildQuerySourcePipeline(firestore, pipelineSource, source);
      break;
    default:
      throw new Error('pipelineExecute() received an unknown source type.');
  }
  if (!currentPipeline) {
    throw new Error(`pipelineExecute() failed to construct source stage "${String(source.source)}" for web SDK.`);
  }
  return currentPipeline;
}
export function buildWebSdkPipeline(firestore, request) {
  const pipelineFactory = firestore.pipeline;
  if (typeof pipelineFactory !== 'function') {
    throw createPipelineRuntimeImportError('pipelineExecute() expected a Firestore instance with pipeline() support.');
  }
  const stack = [{
    kind: 'pipeline',
    pipeline: request.pipeline,
    currentPipeline: buildSourcePipeline(firestore, pipelineFactory, request.pipeline),
    stageIndex: 0
  }];
  let lastCompletedPipeline;
  while (stack.length > 0) {
    const frame = stack[stack.length - 1];
    if (frame.kind === 'union') {
      if (!lastCompletedPipeline) {
        throw new Error('pipelineExecute() failed to rebuild nested union pipeline for web SDK.');
      }
      frame.parent.currentPipeline = frame.method.call(frame.parent.currentPipeline, {
        other: lastCompletedPipeline
      });
      frame.parent.stageIndex += 1;
      lastCompletedPipeline = undefined;
      stack.pop();
      continue;
    }
    if (frame.stageIndex >= frame.pipeline.stages.length) {
      lastCompletedPipeline = frame.currentPipeline;
      stack.pop();
      continue;
    }
    const stage = frame.pipeline.stages[frame.stageIndex];
    if (stage.stage !== 'union') {
      frame.currentPipeline = applyPipelineStage(frame.currentPipeline, stage);
      frame.stageIndex += 1;
      continue;
    }
    const unionPipeline = stage.options.other;
    stack.push({
      kind: 'union',
      parent: frame,
      stage,
      method: getPipelineStageMethod(frame.currentPipeline, stage.stage)
    });
    stack.push({
      kind: 'pipeline',
      pipeline: unionPipeline,
      currentPipeline: buildSourcePipeline(firestore, pipelineFactory, unionPipeline),
      stageIndex: 0
    });
  }
  if (!lastCompletedPipeline) {
    throw new Error('pipelineExecute() failed to rebuild a pipeline for web SDK.');
  }
  return lastCompletedPipeline;
}
