// @react-native-firebase/firestore/dist/module/pipelines/pipeline_support.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

export const PIPELINE_SOURCE_TYPES = ['collection', 'collectionGroup', 'database', 'documents', 'query'];
export const PIPELINE_STAGE_TYPES = ['where', 'select', 'addFields', 'removeFields', 'sort', 'limit', 'offset', 'aggregate', 'distinct', 'findNearest', 'replaceWith', 'sample', 'union', 'unnest', 'rawStage'];
export const PIPELINE_UNSUPPORTED_BASE_MESSAGE = 'Some Firestore pipeline features are not supported by this native implementation yet.';

// Keep this in sync with iOS native support in
// `RNFBFirestorePipelineNodeBuilder.swift`.
// Remove entries once the iOS node builder/runtime path supports them.
const IOS_UNSUPPORTED_FUNCTION_NAMES = new Set(['arrayGet', 'conditional', 'round', 'stringRepeat', 'substring', 'timestampAdd', 'timestampSubtract', 'trunc']);
export function createPipelineUnsupportedMessage(pipeline) {
  const firstStage = pipeline?.stages?.[0];
  if (firstStage?.stage && typeof firstStage.stage === 'string') {
    return `${PIPELINE_UNSUPPORTED_BASE_MESSAGE} Unsupported stage: ${firstStage.stage}.`;
  }
  const source = pipeline?.source?.source;
  if (source && typeof source === 'string') {
    return `${PIPELINE_UNSUPPORTED_BASE_MESSAGE} Unsupported source: ${source}.`;
  }
  return PIPELINE_UNSUPPORTED_BASE_MESSAGE;
}
export function getIOSUnsupportedPipelineFunctions(pipeline) {
  if (!pipeline) {
    return [];
  }
  const unsupported = new Set();
  collectIOSUnsupportedFunctions(pipeline, unsupported);
  return Array.from(unsupported).sort();
}
function collectIOSUnsupportedFunctions(value, unsupported) {
  if (Array.isArray(value)) {
    value.forEach(entry => collectIOSUnsupportedFunctions(entry, unsupported));
    return;
  }
  if (!value || typeof value !== 'object') {
    return;
  }
  const objectValue = value;
  if (isSerializedFunctionExpression(objectValue)) {
    if (IOS_UNSUPPORTED_FUNCTION_NAMES.has(objectValue.name)) {
      unsupported.add(objectValue.name);
    }
    if (Array.isArray(objectValue.args)) {
      objectValue.args.forEach(entry => collectIOSUnsupportedFunctions(entry, unsupported));
    }
  }
  Object.values(objectValue).forEach(entry => {
    collectIOSUnsupportedFunctions(entry, unsupported);
  });
}
function isSerializedFunctionExpression(value) {
  return typeof value.name === 'string' && (value.exprType === 'Function' || value.__kind === 'expression');
}
