// @react-native-firebase/firestore/dist/module/pipelines/pipeline_runtime.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { isArray, isIOS, isNumber, isObject, isString, isUndefined } from '@react-native-firebase/app/dist/module/common';
import FirestorePath from "../FirestorePath.js";
import FirestoreTimestamp from "../FirestoreTimestamp.js";
import DocumentReferenceClass from "../FirestoreDocumentReference.js";
import FieldPath, { fromDotSeparatedString } from "../FieldPath.js";
import { extractFieldPathData } from "../utils/index.js";
import { parseNativeMap } from "../utils/serialize.js";
import { getFirestore } from "../modular.js";
import { getIOSUnsupportedPipelineFunctions } from "./pipeline_support.js";
import { validateExecuteOptions, validateSerializedPipeline } from "./pipeline_validate.js";
const PIPELINE_RUNTIME_SYMBOL = Symbol.for('RNFBFirestorePipelineRuntime');
const PIPELINE_RUNTIME_INSTALLER_SYMBOL = Symbol.for('RNFBFirestorePipelineRuntimeInstaller');
function isRuntimePipeline(value) {
  return isObject(value) && value[PIPELINE_RUNTIME_SYMBOL] === true;
}
function isRecord(value) {
  return isObject(value) && !isArray(value);
}
function hasAnyKey(value, keys) {
  return keys.some(key => Object.prototype.hasOwnProperty.call(value, key));
}
function assertSameFirestoreInstance(actual, expected, method) {
  if (actual !== expected) {
    throw new Error(`firebase.firestore().pipeline().${method}(*) cannot use a reference from a different Firestore instance.`);
  }
}
function getFirestoreReference(value) {
  if (isRecord(value) && isRecord(value.firestore) && isString(value.path) && value.path.length > 0) {
    return {
      firestore: value.firestore,
      path: value.path
    };
  }
  return undefined;
}
function parseTimestamp(value) {
  if (value instanceof FirestoreTimestamp) {
    return value;
  }
  if (isNumber(value)) {
    return FirestoreTimestamp.fromMillis(value);
  }
  if (isArray(value)) {
    const seconds = Number(value[0] ?? 0);
    const nanoseconds = Number(value[1] ?? 0);
    return new FirestoreTimestamp(seconds, nanoseconds);
  }
  if (isRecord(value)) {
    const seconds = Number(value.seconds ?? 0);
    const nanoseconds = Number(value.nanoseconds ?? 0);
    return new FirestoreTimestamp(seconds, nanoseconds);
  }
  return undefined;
}
function getNativeTypeMapArray(value) {
  if (isArray(value) && value.length > 0 && isNumber(value[0])) {
    return value;
  }
  if (isRecord(value) && isNumber(value[0])) {
    return [value[0], value[1]];
  }
  return undefined;
}
function isNativeTypeMapValue(value) {
  return getNativeTypeMapArray(value) !== undefined;
}
function isNativeTypeMapObject(value) {
  if (!isRecord(value)) {
    return false;
  }
  const values = Object.values(value);
  if (values.length === 0) {
    return true;
  }
  return values.every(isNativeTypeMapValue);
}
function fieldPathToSegments(fieldPath) {
  if (isString(fieldPath)) {
    return fromDotSeparatedString(fieldPath)._segments;
  }
  if (fieldPath instanceof FieldPath) {
    return fieldPath._segments;
  }
  if (isRecord(fieldPath) && isString(fieldPath.path)) {
    return fromDotSeparatedString(fieldPath.path)._segments;
  }
  throw new Error("firebase.firestore().pipeline().execute().result.get(*) 'fieldPath' must be a string, FieldPath, or field().");
}
class RuntimePipelineResult {
  constructor(firestore, nativeResult) {
    if (isNativeTypeMapObject(nativeResult.data)) {
      const normalizedData = Object.fromEntries(Object.entries(nativeResult.data ?? {}).map(([key, value]) => [key, getNativeTypeMapArray(value) ?? value]));
      this._data = parseNativeMap(firestore, normalizedData);
    } else {
      this._data = nativeResult.data ?? {};
    }
    if (isString(nativeResult.path)) {
      this.ref = new DocumentReferenceClass(firestore, FirestorePath.fromName(nativeResult.path));
    }
    this.id = isString(nativeResult.id) ? nativeResult.id : this.ref?.id;
    this.createTime = parseTimestamp(nativeResult.createTime);
    this.updateTime = parseTimestamp(nativeResult.updateTime);
  }
  data() {
    return this._data;
  }
  get(fieldPath) {
    return extractFieldPathData(this._data, fieldPathToSegments(fieldPath));
  }
}
function serializeValue(value, visiting) {
  if (value instanceof FirestoreTimestamp) {
    return {
      seconds: value.seconds,
      nanoseconds: value.nanoseconds
    };
  }
  if (value instanceof FieldPath) {
    return {
      segments: value._toArray()
    };
  }
  if (isRuntimePipeline(value)) {
    return value.serialize(visiting);
  }
  if (isArray(value)) {
    return value.map(entry => serializeValue(entry, visiting));
  }
  if (isRecord(value)) {
    if (visiting.has(value)) {
      throw new Error('firebase.firestore().pipeline() failed to serialize arguments because of a circular value.');
    }

    // Explicit wire format for aliased expressions so native (Android) always receives expr, alias, as.
    const rec = value;
    if (rec.__kind === 'aliasedExpression' && rec.expr !== undefined && rec.alias !== undefined) {
      visiting.add(value);
      const inner = rec.expr;
      // Flat form for simple field path: { path, alias, as } so native does not need to parse nested expr.
      if (inner && typeof inner === 'object' && (inner.__kind === 'expression' || inner.exprType === 'Field') && typeof inner.path === 'string' && inner.path.length > 0) {
        visiting.delete(value);
        return {
          path: inner.path,
          alias: rec.alias,
          as: rec.alias
        };
      }
      const result = {
        expr: serializeValue(rec.expr, visiting),
        alias: rec.alias,
        as: rec.alias
      };
      visiting.delete(value);
      return result;
    }

    // Convert document/collection references to a stable path representation.
    if (isString(value.path) && isObject(value.firestore)) {
      return {
        path: value.path
      };
    }
    visiting.add(value);
    const output = {};
    const entries = Object.entries(value);
    for (const [key, entryValue] of entries) {
      if (!isUndefined(entryValue)) {
        output[key] = serializeValue(entryValue, visiting);
      }
    }
    visiting.delete(value);
    // Ensure aliased expressions have both "alias" and "as" for native (e.g. Android coerceSelectable).
    if (typeof output.alias !== 'undefined' && (typeof output.expr !== 'undefined' || output.__kind === 'aliasedExpression')) {
      output.as = output.alias;
    }
    return output;
  }
  return value;
}
function ensureNonEmptyString(path, method) {
  if (!isString(path) || path.length === 0) {
    throw new Error(`firebase.firestore().pipeline().${method}(*) expected a non-empty string.`);
  }
  return path;
}
function asQueryInternals(query) {
  return query;
}
class RuntimePipelineImpl {
  [PIPELINE_RUNTIME_SYMBOL] = true;
  constructor(firestore, source, stages = []) {
    this.firestore = firestore;
    this._source = source;
    this._stages = stages;
  }
  append(stage, options) {
    return new RuntimePipelineImpl(this.firestore, this._source, [...this._stages, {
      stage,
      options
    }]);
  }
  serialize(visiting = new WeakSet()) {
    if (visiting.has(this)) {
      throw new Error('firebase.firestore().pipeline() cannot union a pipeline with itself.');
    }
    visiting.add(this);
    const stages = this._stages.map(stage => ({
      stage: stage.stage,
      options: serializeValue(stage.options, visiting)
    }));
    const source = serializeValue(this._source, visiting);
    visiting.delete(this);
    return {
      source,
      stages
    };
  }
  where(conditionOrOptions) {
    const condition = isRecord(conditionOrOptions) && hasAnyKey(conditionOrOptions, ['condition']) ? conditionOrOptions.condition : conditionOrOptions;
    return this.append('where', {
      condition
    });
  }
  select(...selectionOrOptions) {
    const first = selectionOrOptions[0];
    const selections = selectionOrOptions.length === 1 && isRecord(first) && hasAnyKey(first, ['selection', 'selections']) ? (first.selection ?? first.selections) || [] : selectionOrOptions;
    return this.append('select', {
      selections
    });
  }
  addFields(...fieldOrOptions) {
    const first = fieldOrOptions[0];
    const fields = fieldOrOptions.length === 1 && isRecord(first) && hasAnyKey(first, ['field', 'fields']) ? (first.field ?? first.fields) || [] : fieldOrOptions;
    return this.append('addFields', {
      fields
    });
  }
  removeFields(...fieldOrOptions) {
    const first = fieldOrOptions[0];
    const fields = fieldOrOptions.length === 1 && isRecord(first) && hasAnyKey(first, ['field', 'fields']) ? (first.field ?? first.fields) || [] : fieldOrOptions;
    return this.append('removeFields', {
      fields
    });
  }
  sort(...orderingOrOptions) {
    const first = orderingOrOptions[0];
    const orderings = orderingOrOptions.length === 1 && isRecord(first) && hasAnyKey(first, ['ordering', 'orderings']) ? (first.ordering ?? first.orderings) || [] : orderingOrOptions;
    return this.append('sort', {
      orderings
    });
  }
  limit(nOrOptions) {
    const limit = isNumber(nOrOptions) ? nOrOptions : Number(nOrOptions.n ?? nOrOptions.limit ?? 0);
    return this.append('limit', {
      limit
    });
  }
  offset(nOrOptions) {
    const offset = isNumber(nOrOptions) ? nOrOptions : Number(nOrOptions.n ?? nOrOptions.offset ?? 0);
    return this.append('offset', {
      offset
    });
  }
  aggregate(...accumulatorOrOptions) {
    const first = accumulatorOrOptions[0];
    const isOptionsRecord = accumulatorOrOptions.length === 1 && isRecord(first) && hasAnyKey(first, ['accumulators', 'accumulator', 'groups', 'group']);
    const accumulators = isOptionsRecord ? (first.accumulators ?? first.accumulator) || [] : accumulatorOrOptions;
    const groups = isOptionsRecord ? first.groups ?? first.group : undefined;
    return this.append('aggregate', {
      accumulators,
      groups
    });
  }
  distinct(...groupOrOptions) {
    const first = groupOrOptions[0];
    const groups = groupOrOptions.length === 1 && isRecord(first) && hasAnyKey(first, ['groups', 'group']) ? (first.groups ?? first.group) || [] : groupOrOptions;
    return this.append('distinct', {
      groups
    });
  }
  findNearest(options) {
    return this.append('findNearest', options);
  }
  replaceWith(fieldNameOrExprOrOptions) {
    if (isString(fieldNameOrExprOrOptions)) {
      return this.append('replaceWith', {
        map: fieldNameOrExprOrOptions
      });
    }
    if (isRecord(fieldNameOrExprOrOptions) && hasAnyKey(fieldNameOrExprOrOptions, ['map', 'expr', 'fieldName'])) {
      const map = fieldNameOrExprOrOptions.map ?? fieldNameOrExprOrOptions.expr ?? fieldNameOrExprOrOptions.fieldName;
      return this.append('replaceWith', {
        map
      });
    }
    return this.append('replaceWith', {
      map: fieldNameOrExprOrOptions
    });
  }
  sample(nOrOptions) {
    if (isNumber(nOrOptions)) {
      return this.append('sample', {
        documents: nOrOptions
      });
    }
    return this.append('sample', nOrOptions);
  }
  union(otherOrOptions) {
    const other = isRuntimePipeline(otherOrOptions) ? otherOrOptions : otherOrOptions.pipeline;
    if (!other || !isRuntimePipeline(other)) {
      throw new Error('firebase.firestore().pipeline().union(*) expected a pipeline created from firestore.pipeline().');
    }
    if (other.firestore !== this.firestore) {
      throw new Error('firebase.firestore().pipeline().union(*) cannot combine pipelines from different Firestore instances.');
    }
    return this.append('union', {
      other
    });
  }
  unnest(selectableOrOptions, indexField) {
    const isOptionsObject = isRecord(selectableOrOptions) && (Object.prototype.hasOwnProperty.call(selectableOrOptions, 'indexField') || Object.prototype.hasOwnProperty.call(selectableOrOptions, 'selectable') && !hasAnyKey(selectableOrOptions, ['__kind', 'exprType', 'path', 'name', 'args', 'alias']));
    if (isOptionsObject) {
      return this.append('unnest', selectableOrOptions);
    }
    return this.append('unnest', {
      selectable: selectableOrOptions,
      indexField
    });
  }
  rawStage(name, params, options) {
    return this.append('rawStage', {
      name,
      params,
      options: options ?? {}
    });
  }
}
class RuntimePipelineSourceImpl {
  constructor(firestore) {
    this._firestore = firestore;
  }
  collection(pathOrCollectionRefOrOptions) {
    if (isString(pathOrCollectionRefOrOptions)) {
      return new RuntimePipelineImpl(this._firestore, {
        source: 'collection',
        path: ensureNonEmptyString(pathOrCollectionRefOrOptions, 'collection')
      });
    }
    if (!isRecord(pathOrCollectionRefOrOptions)) {
      throw new Error('firebase.firestore().pipeline().collection(*) expected a path, CollectionReference, or options object.');
    }
    const collectionRef = getFirestoreReference(pathOrCollectionRefOrOptions.collectionRef) ?? getFirestoreReference(pathOrCollectionRefOrOptions);
    if (collectionRef) {
      assertSameFirestoreInstance(collectionRef.firestore, this._firestore, 'collection');
    }
    const path = isString(pathOrCollectionRefOrOptions.path) ? pathOrCollectionRefOrOptions.path : isRecord(pathOrCollectionRefOrOptions.collectionRef) && isString(pathOrCollectionRefOrOptions.collectionRef.path) ? pathOrCollectionRefOrOptions.collectionRef.path : undefined;
    return new RuntimePipelineImpl(this._firestore, {
      source: 'collection',
      path: ensureNonEmptyString(path, 'collection'),
      rawOptions: isRecord(pathOrCollectionRefOrOptions.rawOptions) ? pathOrCollectionRefOrOptions.rawOptions : undefined
    });
  }
  collectionGroup(collectionIdOrOptions) {
    if (isString(collectionIdOrOptions)) {
      return new RuntimePipelineImpl(this._firestore, {
        source: 'collectionGroup',
        collectionId: ensureNonEmptyString(collectionIdOrOptions, 'collectionGroup')
      });
    }
    if (!isRecord(collectionIdOrOptions)) {
      throw new Error('firebase.firestore().pipeline().collectionGroup(*) expected a collectionId string or options object.');
    }
    return new RuntimePipelineImpl(this._firestore, {
      source: 'collectionGroup',
      collectionId: ensureNonEmptyString(collectionIdOrOptions.collectionId, 'collectionGroup'),
      rawOptions: isRecord(collectionIdOrOptions.rawOptions) ? collectionIdOrOptions.rawOptions : undefined
    });
  }
  database(options) {
    return new RuntimePipelineImpl(this._firestore, {
      source: 'database',
      rawOptions: isRecord(options?.rawOptions) ? options.rawOptions : undefined
    });
  }
  documents(docsOrOptions) {
    const docs = isArray(docsOrOptions) ? docsOrOptions : (docsOrOptions.docs ?? docsOrOptions.documents) || [];
    if (!isArray(docs) || docs.length === 0) {
      throw new Error('firebase.firestore().pipeline().documents(*) expected at least one document path or DocumentReference.');
    }
    const documents = docs.map((doc, index) => {
      if (isString(doc) && doc.length > 0) {
        return doc;
      }
      if (isRecord(doc) && isString(doc.path) && doc.path.length > 0) {
        const documentRef = getFirestoreReference(doc);
        if (documentRef) {
          assertSameFirestoreInstance(documentRef.firestore, this._firestore, 'documents');
        }
        return doc.path;
      }
      throw new Error(`firebase.firestore().pipeline().documents(*) invalid value at index ${index}. Expected a document path or DocumentReference.`);
    });
    return new RuntimePipelineImpl(this._firestore, {
      source: 'documents',
      documents
    });
  }
  createFrom(query) {
    const internals = asQueryInternals(query);
    if (!internals || !internals._collectionPath || !internals._modifiers) {
      throw new Error('firebase.firestore().pipeline().createFrom(*) expected a Query from @react-native-firebase/firestore.');
    }
    if (internals._firestore !== this._firestore) {
      throw new Error('firebase.firestore().pipeline().createFrom(*) cannot use a Query from a different Firestore instance.');
    }
    return new RuntimePipelineImpl(this._firestore, {
      source: 'query',
      path: internals._collectionPath.relativeName,
      queryType: internals._modifiers.type,
      filters: internals._modifiers.filters,
      orders: internals._modifiers.orders,
      options: internals._modifiers.options
    });
  }
}
export function createPipelineSource(firestore) {
  return new RuntimePipelineSourceImpl(firestore);
}
export function installPipelineRuntime(firestoreInstance) {
  let firestore = firestoreInstance;
  if (!firestore) {
    try {
      firestore = getFirestore();
    } catch {
      return;
    }
  }
  const prototype = Object.getPrototypeOf(firestore);
  if (!prototype || prototype.__rnfbFirestorePipelineInstalled__) {
    return;
  }
  Object.defineProperty(prototype, 'pipeline', {
    value() {
      return createPipelineSource(this);
    },
    configurable: true,
    enumerable: false,
    writable: true
  });
  Object.defineProperty(prototype, '__rnfbFirestorePipelineInstalled__', {
    value: true,
    configurable: true,
    enumerable: false,
    writable: true
  });
}
export function registerPipelineRuntimeInstaller() {
  const runtimeGlobal = globalThis;
  runtimeGlobal[PIPELINE_RUNTIME_INSTALLER_SYMBOL] = installPipelineRuntime;
}
function parseExecuteInput(pipelineOrOptions) {
  if (isRuntimePipeline(pipelineOrOptions)) {
    return {
      runtimePipeline: pipelineOrOptions,
      executeOptions: {}
    };
  }
  if (!isRecord(pipelineOrOptions)) {
    throw new Error('firebase.firestore().pipeline().execute(*) expected a Pipeline or PipelineExecuteOptions.');
  }
  const runtimePipeline = pipelineOrOptions.pipeline;
  if (!isRuntimePipeline(runtimePipeline)) {
    throw new Error('firebase.firestore().pipeline().execute(*) expected options.pipeline to be created from firestore.pipeline().');
  }
  return {
    runtimePipeline,
    executeOptions: validateExecuteOptions({
      indexMode: pipelineOrOptions.indexMode,
      rawOptions: pipelineOrOptions.rawOptions
    }) ?? {}
  };
}
export async function executeRuntimePipeline(pipelineOrOptions) {
  const {
    runtimePipeline,
    executeOptions
  } = parseExecuteInput(pipelineOrOptions);
  const serializedPipeline = runtimePipeline.serialize();
  validateSerializedPipeline(serializedPipeline);
  if (isIOS) {
    const unsupportedFunctions = getIOSUnsupportedPipelineFunctions(serializedPipeline);
    if (unsupportedFunctions.length) {
      throw new Error(`pipelineExecute() does not support these functions on iOS yet: ${unsupportedFunctions.join(', ')}.`);
    }
  }
  const nativeResponse = await runtimePipeline.firestore.native.pipelineExecute(serializedPipeline, executeOptions);
  const executionTime = parseTimestamp(nativeResponse?.executionTime);
  if (!executionTime) {
    throw new Error('firebase.firestore().pipeline().execute(*) expected pipelineExecute() to return executionTime.');
  }
  const results = (nativeResponse?.results ?? []).map(result => new RuntimePipelineResult(runtimePipeline.firestore, result));
  return {
    results,
    executionTime
  };
}
