// @react-native-firebase/firestore/dist/module/modular/Bytes.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import Blob from "../FirestoreBlob.js";
export class Bytes extends Blob {
  constructor(blob) {
    super(true);
    // Binary string was already parsed and created; avoid reparsing.
    this._binaryString = blob._binaryString;
  }
  static fromBase64String(base64) {
    try {
      return new Bytes(Blob.fromBase64String(base64));
    } catch (error) {
      throw new Error(`Failed to construct data from Base64 string: ${error}`);
    }
  }
  static fromUint8Array(array) {
    return new Bytes(Blob.fromUint8Array(array));
  }
  toBase64() {
    return super.toBase64();
  }
  toUint8Array() {
    return super.toUint8Array();
  }
  toString() {
    return `Bytes(base64: ${this.toBase64()})`;
  }
  isEqual(other) {
    return super.isEqual(other);
  }
  static _jsonSchemaVersion = 'firestore/bytes/1.0';
  static _jsonSchema = {
    type: Bytes._jsonSchemaVersion,
    bytes: 'string'
  };
  toJSON() {
    return {
      type: Bytes._jsonSchemaVersion,
      bytes: this.toBase64()
    };
  }
  static fromJSON(json) {
    if (json && typeof json.type === 'string' && json.type === Bytes._jsonSchemaVersion && typeof json.bytes === 'string') {
      return Bytes.fromBase64String(json.bytes);
    }
    throw new Error('Unexpected error creating Bytes from JSON.');
  }
}
