// @react-native-firebase/firestore/dist/module/modular/query.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { MODULAR_DEPRECATION_ARG } from '@react-native-firebase/app/dist/module/common';
import { Filter } from "../FirestoreFilter.js";
/**
 * Abstraction of a constraint that can be applied to a Firestore query.
 * Not exported — public API matches firebase-js-sdk ({@link QueryConstraint} only).
 */
class AppliableConstraint {}

/**
 * A `QueryConstraint` is used to narrow the set of documents returned by a
 * Firestore query. `QueryConstraint`s are created by invoking {@link where},
 * {@link orderBy}, {@link startAt}, {@link startAfter}, {@link endBefore},
 * {@link endAt}, {@link limit}, {@link limitToLast} and can then be passed to
 * {@link query} to create a new query instance that also contains this `QueryConstraint`.
 */
export class QueryConstraint extends AppliableConstraint {
  _apply(_queryRef) {
    throw new Error('_apply must be implemented by subclass');
  }
}

/**
 * Shared _apply for orderBy/limit/startAt/endAt/where. Not exported — implementation detail.
 */
class QueryConstraintBase extends QueryConstraint {
  constructor(...args) {
    super();
    this._args = args;
  }
  _apply(query) {
    const method = query[this.type];
    if (!method) {
      throw new Error(`Query method '${this.type}' is not available on query instance.`);
    }
    return method.call(query, ...this._args, MODULAR_DEPRECATION_ARG);
  }
}
export class QueryCompositeFilterConstraint extends AppliableConstraint {
  /**
   * @internal
   */
  constructor(/** The type of this query constraint */
  type, _queryConstraints) {
    super();
    this.type = type;
    this._queryConstraints = _queryConstraints;
  }
  static _create(type, _queryConstraints) {
    // Validate nested OR filters when creating the constraint
    if (type === 'or') {
      const filters = _queryConstraints.map(constraint => {
        if (constraint instanceof QueryCompositeFilterConstraint) {
          return constraint._filter;
        }
        return constraint._filter;
      });
      // This will throw if nested OR filters are detected
      Filter.or(...filters);
    }
    return new QueryCompositeFilterConstraint(type, _queryConstraints);
  }
  _apply(query) {
    const filters = this._queryConstraints.map(constraint => {
      if (constraint instanceof QueryCompositeFilterConstraint) {
        return constraint._filter;
      }
      return constraint._filter;
    });
    const _filter = this.type === 'or' ? Filter.or(...filters) : Filter.and(...filters);
    const where = query.where;
    return where.call(query, _filter, MODULAR_DEPRECATION_ARG);
  }
  get _filter() {
    const filters = this._queryConstraints.map(constraint => {
      if (constraint instanceof QueryCompositeFilterConstraint) {
        return constraint._filter;
      }
      return constraint._filter;
    });
    return this.type === 'or' ? Filter.or(...filters) : Filter.and(...filters);
  }
}
export class QueryOrderByConstraint extends QueryConstraintBase {
  type = 'orderBy';
  constructor(fieldPath, directionStr) {
    super(fieldPath, directionStr);
  }
}
export class QueryLimitConstraint extends QueryConstraintBase {
  constructor(type, limitValue) {
    super(limitValue);
    this.type = type;
  }
}
export class QueryStartAtConstraint extends QueryConstraintBase {
  constructor(type, ...docOrFields) {
    super(...docOrFields);
    this.type = type;
  }
}
export class QueryEndAtConstraint extends QueryConstraintBase {
  constructor(type, ...fieldValues) {
    super(...fieldValues);
    this.type = type;
  }
}
export class QueryFieldFilterConstraint extends QueryConstraintBase {
  type = 'where';
  constructor(fieldPath, opStr, value) {
    super(fieldPath, opStr, value);
    this._filter = Filter(fieldPath, opStr, value);
  }
}
export function query(queryRef, queryConstraint, ...additionalQueryConstraints) {
  let queryConstraints = [];
  if (queryConstraint instanceof AppliableConstraint) {
    queryConstraints.push(queryConstraint);
  }
  queryConstraints = queryConstraints.concat(additionalQueryConstraints);
  let constrainedQuery = queryRef;
  for (const constraint of queryConstraints) {
    if (!constraint) {
      continue;
    }
    constrainedQuery = constraint._apply(constrainedQuery);
  }
  return constrainedQuery;
}
export function where(fieldPath, opStr, value) {
  return new QueryFieldFilterConstraint(fieldPath, opStr, value);
}
export function or(...queryConstraints) {
  return QueryCompositeFilterConstraint._create('or', queryConstraints);
}
export function and(...queryConstraints) {
  return QueryCompositeFilterConstraint._create('and', queryConstraints);
}
export function orderBy(fieldPath, directionStr) {
  return new QueryOrderByConstraint(fieldPath, directionStr);
}
export function startAt(...docOrFields) {
  return new QueryStartAtConstraint('startAt', ...docOrFields);
}
export function startAfter(...docOrFields) {
  return new QueryStartAtConstraint('startAfter', ...docOrFields);
}
export function endAt(...args) {
  return new QueryEndAtConstraint('endAt', ...args);
}
export function endBefore(...fieldValues) {
  return new QueryEndAtConstraint('endBefore', ...fieldValues);
}
export function limit(limitValue) {
  return new QueryLimitConstraint('limit', limitValue);
}
export function limitToLast(limitValue) {
  return new QueryLimitConstraint('limitToLast', limitValue);
}
export function getDoc(reference) {
  const get = reference.get;
  return get.call(reference, {
    source: 'default'
  }, MODULAR_DEPRECATION_ARG);
}
export function getDocFromCache(reference) {
  const get = reference.get;
  return get.call(reference, {
    source: 'cache'
  }, MODULAR_DEPRECATION_ARG);
}
export function getDocFromServer(reference) {
  const get = reference.get;
  return get.call(reference, {
    source: 'server'
  }, MODULAR_DEPRECATION_ARG);
}
export function getDocs(queryRef) {
  const get = queryRef.get;
  return get.call(queryRef, {
    source: 'default'
  }, MODULAR_DEPRECATION_ARG);
}
export function getDocsFromCache(queryRef) {
  const get = queryRef.get;
  return get.call(queryRef, {
    source: 'cache'
  }, MODULAR_DEPRECATION_ARG);
}
export function getDocsFromServer(queryRef) {
  const get = queryRef.get;
  return get.call(queryRef, {
    source: 'server'
  }, MODULAR_DEPRECATION_ARG);
}
export function deleteDoc(reference) {
  const remove = reference.delete;
  return remove.call(reference, MODULAR_DEPRECATION_ARG);
}
export function queryEqual(left, right) {
  const isEqual = left.isEqual;
  return isEqual.call(left, right, MODULAR_DEPRECATION_ARG);
}
