// @react-native-firebase/firestore/dist/module/utils/serialize.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import { isArray, isBoolean, isDate, isNull, isNumber, isObject, isString, isUndefined } from '@react-native-firebase/app/dist/module/common';
import Blob from "../FirestoreBlob.js";
import { DOCUMENT_ID } from "../FieldPath.js";
import FirestoreGeoPoint from "../FirestoreGeoPoint.js";
import FirestorePath from "../FirestorePath.js";
import FirestoreTimestamp from "../FirestoreTimestamp.js";
import { getTypeMapInt, getTypeMapName } from "./typemap.js";
import { Bytes } from "../modular/Bytes.js";
import FirestoreVectorValue from "../FirestoreVectorValue.js";
let FirestoreDocumentReference = null;
export function provideDocumentReferenceClass(documentReference) {
  FirestoreDocumentReference = documentReference;
}
let FirestoreFieldValue = null;
export function provideFieldValueClass(fieldValue) {
  FirestoreFieldValue = fieldValue;
}
export function buildNativeMap(data, ignoreUndefined) {
  const nativeData = {};
  if (data) {
    const keys = Object.keys(data);
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i];
      if (typeof data[key] === 'undefined') {
        if (!ignoreUndefined) {
          throw new Error('Unsupported field value: undefined');
        }
        continue;
      }
      const typeMap = generateNativeData(data[key], ignoreUndefined);
      if (typeMap) {
        nativeData[key] = typeMap;
      }
    }
  }
  return nativeData;
}
export function buildNativeArray(array, ignoreUndefined) {
  const nativeArray = [];
  if (array) {
    for (let i = 0; i < array.length; i++) {
      const value = array[i];
      if (typeof value === 'undefined') {
        if (!ignoreUndefined) {
          throw new Error('Unsupported field value: undefined');
        }
        continue;
      }
      const typeMap = generateNativeData(value, ignoreUndefined);
      if (typeMap) {
        nativeArray.push(typeMap);
      }
    }
  }
  return nativeArray;
}
export function generateNativeData(value, ignoreUndefined) {
  if (Number.isNaN(value)) {
    return getTypeMapInt('nan');
  }
  if (value === Number.NEGATIVE_INFINITY) {
    return getTypeMapInt('-infinity');
  }
  if (value === Number.POSITIVE_INFINITY) {
    return getTypeMapInt('infinity');
  }
  if (isNull(value) || isUndefined(value)) {
    return getTypeMapInt('null');
  }
  if (value === DOCUMENT_ID) {
    return getTypeMapInt('documentid');
  }
  if (isBoolean(value)) {
    return value ? getTypeMapInt('booleanTrue') : getTypeMapInt('booleanFalse');
  }
  if (isNumber(value)) {
    if (value === 0 && 1 / value === -Infinity) {
      return getTypeMapInt('negativeZero');
    }
    if (Number.isSafeInteger(value)) {
      return getTypeMapInt('integer', value);
    }
    return getTypeMapInt('double', value);
  }
  if (isString(value)) {
    return value === '' ? getTypeMapInt('stringEmpty') : getTypeMapInt('string', value);
  }
  if (isArray(value)) {
    return getTypeMapInt('array', buildNativeArray(value, ignoreUndefined));
  }
  if (isObject(value)) {
    if (FirestoreDocumentReference && value instanceof FirestoreDocumentReference) {
      return getTypeMapInt('reference', value.path);
    }
    if (value instanceof FirestoreGeoPoint) {
      return getTypeMapInt('geopoint', [value.latitude, value.longitude]);
    }
    if (isDate(value)) {
      const timestamp = FirestoreTimestamp.fromDate(value);
      return getTypeMapInt('timestamp', [timestamp.seconds, timestamp.nanoseconds]);
    }
    if (value instanceof FirestoreTimestamp) {
      return getTypeMapInt('timestamp', [value.seconds, value.nanoseconds]);
    }
    if (value instanceof Blob || value instanceof Bytes) {
      return getTypeMapInt('blob', value.toBase64());
    }
    if (FirestoreFieldValue && value instanceof FirestoreFieldValue) {
      return getTypeMapInt('fieldvalue', [value._type, value._elements]);
    }
    if (value instanceof FirestoreVectorValue) {
      return getTypeMapInt('vector', value.toArray());
    }
    return getTypeMapInt('object', buildNativeMap(value, ignoreUndefined));
  }

  // eslint-disable-next-line no-console
  console.warn(`Unknown data type received ${value}`);
  return getTypeMapInt('unknown');
}
function getArrayLikeEntries(value) {
  if (isArray(value)) {
    return value;
  }
  if (isObject(value) && !isArray(value)) {
    const record = value;
    if (!Object.prototype.hasOwnProperty.call(record, '0')) {
      return undefined;
    }
    const entries = [];
    let index = 0;
    while (Object.prototype.hasOwnProperty.call(record, String(index))) {
      entries.push(record[String(index)]);
      index += 1;
    }
    return entries;
  }
  return undefined;
}
export function parseNativeMap(firestore, nativeData) {
  let data;
  if (nativeData) {
    data = {};
    const entries = Object.entries(nativeData);
    for (let i = 0; i < entries.length; i++) {
      const [key, typeArray] = entries[i];
      data[key] = parseNativeData(firestore, typeArray);
    }
  }
  return data;
}
export function parseNativeArray(firestore, nativeArray) {
  const array = [];
  const normalizedArray = getArrayLikeEntries(nativeArray) ?? [];
  if (normalizedArray) {
    for (let i = 0; i < normalizedArray.length; i++) {
      array.push(parseNativeData(firestore, normalizedArray[i]));
    }
  }
  return array;
}
export function parseNativeData(firestore, nativeArray) {
  const normalizedTypeMap = getArrayLikeEntries(nativeArray);
  if (!normalizedTypeMap || !isNumber(normalizedTypeMap[0])) {
    // eslint-disable-next-line no-console
    console.warn('Unknown data type received from native channel: malformed type map');
    return undefined;
  }
  const [int, value] = normalizedTypeMap;
  const type = getTypeMapName(int);
  switch (type) {
    case 'nan':
      return NaN;
    case '-infinity':
      return -Infinity;
    case 'infinity':
      return Infinity;
    case 'null':
      return null;
    case 'booleanTrue':
      return true;
    case 'booleanFalse':
      return false;
    case 'double':
    case 'integer':
    case 'negativeZero':
    case 'string':
      return value;
    case 'stringEmpty':
      return '';
    case 'array':
      return parseNativeArray(firestore, getArrayLikeEntries(value) ?? []);
    case 'object':
      return parseNativeMap(firestore, value);
    case 'reference':
      return new FirestoreDocumentReference(firestore, FirestorePath.fromName(value));
    case 'geopoint':
      {
        const v = getArrayLikeEntries(value) ?? [];
        return new FirestoreGeoPoint(v[0] ?? 0, v[1] ?? 0);
      }
    case 'timestamp':
      {
        const v = getArrayLikeEntries(value) ?? [];
        return new FirestoreTimestamp(v[0] ?? 0, v[1] ?? 0);
      }
    case 'blob':
      return Bytes.fromBase64String(value);
    case 'vector':
      return new FirestoreVectorValue(value);
    default:
      // eslint-disable-next-line no-console
      console.warn(`Unknown data type received from native channel: ${type}`);
      return value;
  }
}
